File: scripts/e2e/homepage_contact_submit.py
#!/usr/bin/env python3
"""End-to-end test: submit the homepage contact form with a valid email
and verify a success toast appears (and the form is reset).

Run:
    python3 scripts/e2e/homepage_contact_submit.py
    python3 scripts/e2e/homepage_contact_submit.py --url http://localhost:8080 --locale en

Exit codes: 0 pass, 1 assertion failure, 2 harness error.

Uses the sandbox's pre-installed Playwright + bundled Chromium. No project
dependency is added; run this against a locally-running dev server.
"""
from __future__ import annotations

import argparse
import asyncio
import re
import sys
import time
from pathlib import Path

try:
    from playwright.async_api import async_playwright
except ImportError:  # pragma: no cover - harness error
    print("Playwright is not available in this environment.", file=sys.stderr)
    sys.exit(2)


SCREENSHOTS = Path("/tmp/browser/homepage-contact-submit")
SCREENSHOTS.mkdir(parents=True, exist_ok=True)

NOISY_CONSOLE = re.compile(r"vite|hydrated|hydration|websocket|hmr", re.IGNORECASE)
FAILURE_TOAST = re.compile(
    r"required|invalid|valid email|please review|error|failed|too many",
    re.IGNORECASE,
)


async def run(url: str, locale: str) -> int:
    failures: list[str] = []

    def check(cond: bool, msg: str) -> None:
        if not cond:
            failures.append(msg)

    async with async_playwright() as pw:
        browser = await pw.chromium.launch(headless=True)
        context = await browser.new_context(viewport={"width": 1280, "height": 1800})
        page = await context.new_page()

        console_errors: list[str] = []
        page.on(
            "console",
            lambda msg: console_errors.append(msg.text) if msg.type == "error" else None,
        )

        await page.goto(f"{url}/{locale}", wait_until="domcontentloaded")
        await page.screenshot(path=str(SCREENSHOTS / "1_loaded.png"))

        # Best-effort cookie banner dismissal.
        try:
            await page.get_by_role("button", name=re.compile(r"decline|accept", re.I)).first.click(
                timeout=1500
            )
        except Exception:
            pass

        contact = page.locator("#contact")
        await contact.scroll_into_view_if_needed()
        form = page.locator("#contact form").first
        await form.wait_for(state="visible", timeout=10_000)

        email = f"qa+{int(time.time() * 1000)}@nevoindustrial.com"
        await form.locator('input[name="name"]').fill("QA Bot")
        await form.locator('input[name="email"]').fill(email)
        await form.locator('textarea[name="message"]').fill(
            "Automated end-to-end verification of the homepage contact form submission path."
        )
        await page.screenshot(path=str(SCREENSHOTS / "2_filled.png"))

        submit = form.get_by_role(
            "button", name=re.compile(r"send|submit|inquiry|contact", re.I)
        ).first
        await submit.click()

        toast_text = ""
        toast = page.locator("[data-sonner-toast]").first
        try:
            await toast.wait_for(state="visible", timeout=10_000)
            toast_text = (await toast.inner_text()).strip()
        except Exception:
            pass

        await page.screenshot(path=str(SCREENSHOTS / "3_after_submit.png"))

        check(
            bool(toast_text),
            f"No toast appeared after submitting the homepage contact form (email={email}).",
        )
        check(
            not FAILURE_TOAST.search(toast_text),
            f"Toast looked like a failure, not a success: {toast_text!r}",
        )

        email_after = await form.locator('input[name="email"]').input_value()
        check(
            email_after == "",
            f"Expected the form to reset after success, but email is still {email_after!r}.",
        )

        meaningful = [e for e in console_errors if not NOISY_CONSOLE.search(e)]
        if meaningful:
            print("Console errors during run:")
            for err in meaningful:
                print(f"  - {err}")

        await browser.close()

    if failures:
        print(f"FAIL homepage-contact-submit ({len(failures)}):", file=sys.stderr)
        for f in failures:
            print(f"  - {f}", file=sys.stderr)
        print(f"Screenshots saved to {SCREENSHOTS}", file=sys.stderr)
        return 1

    print(f"PASS homepage-contact-submit — toast: {toast_text!r}")
    print(f"Screenshots saved to {SCREENSHOTS}")
    return 0


def main() -> int:
    ap = argparse.ArgumentParser(description=__doc__)
    ap.add_argument("--url", default="http://localhost:8080")
    ap.add_argument("--locale", default="en")
    args = ap.parse_args()
    try:
        return asyncio.run(run(args.url, args.locale))
    except Exception as exc:  # pragma: no cover
        print(f"Harness error: {exc!r}", file=sys.stderr)
        return 2


if __name__ == "__main__":
    sys.exit(main())

File: scripts/e2e/homepage-contact-submit.mjs
#!/usr/bin/env node
/**
 * End-to-end test: submit the homepage contact form with a valid email
 * and verify a success toast appears (and the form is reset).
 *
 * Run:
 *   node scripts/e2e/homepage-contact-submit.mjs
 *   node scripts/e2e/homepage-contact-submit.mjs --url=http://localhost:8080
 *
 * Exit codes: 0 pass, 1 assertion failure, 2 harness error.
 */
import { chromium } from "playwright";

const args = new Map(
  process.argv.slice(2).map((a) => {
    const [k, v] = a.replace(/^--/, "").split("=");
    return [k, v ?? "true"];
  }),
);
const URL = args.get("url") ?? "http://localhost:8080";
const LOCALE = args.get("locale") ?? "en";

const failures = [];
function assert(cond, msg) {
  if (!cond) failures.push(msg);
}

let browser;
try {
  browser = await chromium.launch({ headless: true });
  const context = await browser.new_context
    ? await browser.new_context({ viewport: { width: 1280, height: 1800 } })
    : await browser.newContext({ viewport: { width: 1280, height: 1800 } });
  const page = await context.newPage();

  const consoleErrors = [];
  page.on("console", (msg) => {
    if (msg.type() === "error") consoleErrors.push(msg.text());
  });

  await page.goto(`${URL}/${LOCALE}`, { waitUntil: "domcontentloaded" });

  // Dismiss cookie banner if present (best-effort).
  try {
    const decline = page.getByRole("button", { name: /decline|accept/i }).first();
    await decline.click({ timeout: 1500 });
  } catch {
    /* no banner */
  }

  // Scroll the contact form into view.
  await page.locator("#contact").scrollIntoViewIfNeeded();
  const form = page.locator("#contact form").first();
  await form.waitFor({ state: "visible", timeout: 10_000 });

  // Fill the required fields with valid values.
  const email = `qa+${Date.now()}@nevoindustrial.com`;
  await form.locator('input[name="name"]').fill("QA Bot");
  await form.locator('input[name="email"]').fill(email);
  await form
    .locator('textarea[name="message"]')
    .fill("Automated end-to-end verification of the homepage contact form submission path.");

  // Submit and wait for the success toast rendered by Sonner.
  const submit = form.getByRole("button", { name: /send|submit|inquiry|contact/i }).first();
  await submit.click();

  // Sonner renders toasts in an [aria-live] region with data-sonner-toast items.
  const toast = page.locator("[data-sonner-toast]").first();
  let toastText = "";
  try {
    await toast.waitFor({ state: "visible", timeout: 10_000 });
    toastText = (await toast.innerText()).trim();
  } catch {
    // fallthrough — assertion will report the failure below
  }

  assert(
    toastText.length > 0,
    `No toast appeared after submitting the homepage contact form (email=${email}).`,
  );

  // A validation failure would surface a "required" or "valid" message; a
  // success toast uses the localized inquiry-submitted copy. Reject anything
  // that clearly indicates a validation or delivery failure.
  const lower = toastText.toLowerCase();
  assert(
    !/required|invalid|valid email|please review|error|failed|too many/i.test(lower),
    `Toast looked like a failure, not a success: ${JSON.stringify(toastText)}`,
  );

  // The form should reset on success — the email input should now be empty.
  const emailAfter = await form.locator('input[name="email"]').inputValue();
  assert(
    emailAfter === "",
    `Expected the form to reset after success, but email is still ${JSON.stringify(emailAfter)}.`,
  );

  // Report console errors as soft context (do not fail the run for HMR/vite noise).
  const noisy = /vite|hydrated|hydration|websocket|hmr/i;
  const meaningful = consoleErrors.filter((e) => !noisy.test(e));
  if (meaningful.length) {
    console.warn(`Console errors during run:\n  - ${meaningful.join("\n  - ")}`);
  }

  if (failures.length) {
    console.error(`FAIL homepage-contact-submit (${failures.length}):`);
    for (const f of failures) console.error(`  - ${f}`);
    console.error(`Last toast text: ${JSON.stringify(toastText)}`);
    process.exit(1);
  }

  console.log(`PASS homepage-contact-submit — toast: ${JSON.stringify(toastText)}`);
  process.exit(0);
} catch (err) {
  console.error("Harness error:", err?.stack ?? err);
  process.exit(2);
} finally {
  if (browser) await browser.close();
}
